File: src/app/ActorLibrary.cpp
#include "ActorLibrary.h"

#include "core/Paths.h"

#include <QDir>
#include <QFile>
#include <QFileInfo>
#include <QJsonArray>
#include <QJsonDocument>
#include <QJsonObject>
#include <QUuid>

namespace {

constexpr const char *kFileName = "actors.json";
constexpr const char *kPortraitFolder = "actors";

} // namespace

ActorLibrary::ActorLibrary(QObject *parent)
    : QAbstractListModel(parent)
{
    load();
}

QString ActorLibrary::storeFile() const
{
    return QDir(paths::configDir()).filePath(QString::fromLatin1(kFileName));
}

QString ActorLibrary::portraitDir() const
{
    return QDir(paths::configDir()).filePath(QString::fromLatin1(kPortraitFolder));
}

// ---------------------------------------------------------------------------
// Model
// ---------------------------------------------------------------------------
int ActorLibrary::rowCount(const QModelIndex &parent) const
{
    return parent.isValid() ? 0 : m_actors.size();
}

QVariant ActorLibrary::data(const QModelIndex &index, int role) const
{
    if (index.row() < 0 || index.row() >= m_actors.size())
        return {};

    const QVariantMap &actor = m_actors.at(index.row());
    switch (role) {
    case IdRole:       return actor.value(QStringLiteral("id"));
    case NameRole:     return actor.value(QStringLiteral("name"));
    case PortraitRole: return actor.value(QStringLiteral("portraitPath"));
    case BriefRole:    return actor.value(QStringLiteral("brief"));
    case DecorRole:    return actor.value(QStringLiteral("decor"));
    case ActorRole:    return actor;
    default:           return {};
    }
}

QHash<int, QByteArray> ActorLibrary::roleNames() const
{
    return {{IdRole, "actorId"},
            {NameRole, "name"},
            {PortraitRole, "portraitPath"},
            {BriefRole, "brief"},
            {DecorRole, "decor"},
            {ActorRole, "actor"}};
}

// ---------------------------------------------------------------------------
// Storage
// ---------------------------------------------------------------------------
void ActorLibrary::load()
{
    QFile file(storeFile());
    if (!file.open(QIODevice::ReadOnly))
        return;

    const QJsonArray array = QJsonDocument::fromJson(file.readAll()).array();

    beginResetModel();
    m_actors.clear();
    for (const QJsonValue &entry : array) {
        const QVariantMap actor = entry.toObject().toVariantMap();
        if (!actor.value(QStringLiteral("id")).toString().isEmpty())
            m_actors.append(actor);
    }
    endResetModel();

    emit countChanged();
}

void ActorLibrary::persist()
{
    if (paths::ensureDir(paths::configDir()).isEmpty())
        return;

    QJsonArray array;
    for (const QVariantMap &actor : m_actors)
        array.append(QJsonObject::fromVariantMap(actor));

    QFile file(storeFile());
    if (file.open(QIODevice::WriteOnly | QIODevice::Truncate))
        file.write(QJsonDocument(array).toJson(QJsonDocument::Indented));
}

void ActorLibrary::refresh()
{
    load();
}

QString ActorLibrary::adoptPortrait(const QString &id, const QString &path) const
{
    if (path.isEmpty() || !QFileInfo::exists(path))
        return path;

    const QString dir = paths::ensureDir(portraitDir());
    if (dir.isEmpty())
        return path;

    // Already ours: a re-save of an unchanged actor must not copy the file onto
    // itself and must not accumulate duplicates.
    if (QFileInfo(path).absolutePath() == QFileInfo(dir).absoluteFilePath())
        return path;

    const QString extension = QFileInfo(path).suffix().isEmpty()
                                  ? QStringLiteral("png")
                                  : QFileInfo(path).suffix();
    const QString target = QDir(dir).filePath(QStringLiteral("%1.%2").arg(id, extension));

    QFile::remove(target);
    if (!QFile::copy(path, target))
        return path;

    return target;
}

// ---------------------------------------------------------------------------
// Entries
// ---------------------------------------------------------------------------
QString ActorLibrary::save(const QVariantMap &actor)
{
    QVariantMap entry = actor;

    QString id = entry.value(QStringLiteral("id")).toString();
    if (id.isEmpty())
        id = QUuid::createUuid().toString(QUuid::WithoutBraces).left(8);
    entry.insert(QStringLiteral("id"), id);

    if (entry.value(QStringLiteral("name")).toString().trimmed().isEmpty())
        entry.insert(QStringLiteral("name"), suggestedName());

    entry.insert(QStringLiteral("portraitPath"),
                 adoptPortrait(id, entry.value(QStringLiteral("portraitPath")).toString()));

    int existing = -1;
    for (int i = 0; i < m_actors.size(); ++i) {
        if (m_actors.at(i).value(QStringLiteral("id")).toString() == id) {
            existing = i;
            break;
        }
    }

    if (existing >= 0) {
        m_actors[existing] = entry;
        const QModelIndex changed = index(existing);
        emit dataChanged(changed, changed);
    } else {
        // Newest first: the actor just cast is the one about to be used.
        beginInsertRows({}, 0, 0);
        m_actors.prepend(entry);
        endInsertRows();
        emit countChanged();
    }

    persist();
    return id;
}

QVariantMap ActorLibrary::actor(const QString &id) const
{
    for (const QVariantMap &entry : m_actors) {
        if (entry.value(QStringLiteral("id")).toString() == id)
            return entry;
    }
    return {};
}

void ActorLibrary::remove(const QString &id)
{
    for (int i = 0; i < m_actors.size(); ++i) {
        if (m_actors.at(i).value(QStringLiteral("id")).toString() != id)
            continue;

        const QString portrait = m_actors.at(i).value(QStringLiteral("portraitPath")).toString();

        beginRemoveRows({}, i, i);
        m_actors.removeAt(i);
        endRemoveRows();

        // Only files we adopted are ours to delete; a portrait the user pointed
        // at from their own folders stays where it is.
        if (!portrait.isEmpty()
            && QFileInfo(portrait).absolutePath() == QFileInfo(portraitDir()).absoluteFilePath()) {
            QFile::remove(portrait);
        }

        persist();
        emit countChanged();
        return;
    }
}

QString ActorLibrary::suggestedName() const
{
    return tr("Actor %1").arg(m_actors.size() + 1);
}

File: market_queen_flutter_app/lib/models/actor_library.dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter/foundation.dart';
import 'package:path/path.dart' as p;

import '../core/paths.dart';
import '../i18n/translator.dart';

/// The actors the user has kept.
///
/// An actor is the expensive part of a good ad: it takes several batches to
/// land a face that reads as real, and once it does it should never have to be
/// found again. Saving one copies its portrait out of the casting scratch
/// folder into the config directory, so clearing scratch can never orphan a
/// saved actor.
class ActorLibrary extends ChangeNotifier {
  ActorLibrary() {
    _load();
  }

  static const _fileName = 'actors.json';
  static const _portraitFolder = 'actors';

  final List<Map<String, Object?>> _actors = [];

  List<Map<String, Object?>> get actors => List.unmodifiable(_actors);
  int get count => _actors.length;

  String get _storeFile => p.join(Paths.configDir, _fileName);
  String get _portraitDir => p.join(Paths.configDir, _portraitFolder);

  // ---- Storage ----------------------------------------------------------

  void _load() {
    final file = File(_storeFile);
    if (!file.existsSync()) return;

    Object? decoded;
    try {
      decoded = jsonDecode(file.readAsStringSync());
    } on Exception {
      return;
    }
    if (decoded is! List) return;

    _actors
      ..clear()
      ..addAll([
        for (final entry in decoded)
          if (entry is Map && '${entry['id'] ?? ''}'.isNotEmpty)
            entry.cast<String, Object?>(),
      ]);

    notifyListeners();
  }

  void _persist() {
    if (Paths.ensureDir(Paths.configDir).isEmpty) return;
    try {
      File(_storeFile)
          .writeAsStringSync(const JsonEncoder.withIndent('    ').convert(_actors));
    } on FileSystemException {
      // The next save tries again.
    }
  }

  void refresh() {
    _load();
    notifyListeners();
  }

  /// Copies a portrait out of scratch into the library folder, returning the
  /// new path. Files already inside the library are left where they are.
  String _adoptPortrait(String id, String path) {
    if (path.isEmpty || !File(path).existsSync()) return path;

    final dir = Paths.ensureDir(_portraitDir);
    if (dir.isEmpty) return path;

    // Already ours: a re-save of an unchanged actor must not copy the file onto
    // itself and must not accumulate duplicates.
    if (p.equals(p.dirname(File(path).absolute.path), dir)) return path;

    final extension = p.extension(path).replaceFirst('.', '');
    final target = p.join(dir, '$id.${extension.isEmpty ? 'png' : extension}');

    try {
      final existing = File(target);
      if (existing.existsSync()) existing.deleteSync();
      File(path).copySync(target);
      return target;
    } on FileSystemException {
      return path;
    }
  }

  // ---- Entries ----------------------------------------------------------

  /// Inserts or updates, and returns the actor's id. An actor carrying an id
  /// that is already on file is updated in place rather than duplicated.
  String save(Map<String, Object?> actor) {
    final entry = Map<String, Object?>.of(actor);

    var id = '${entry['id'] ?? ''}';
    if (id.isEmpty) id = _newId();
    entry['id'] = id;

    if ('${entry['name'] ?? ''}'.trim().isEmpty) {
      entry['name'] = suggestedName();
    }

    entry['portraitPath'] = _adoptPortrait(id, '${entry['portraitPath'] ?? ''}');

    final existing = _actors.indexWhere((a) => '${a['id'] ?? ''}' == id);
    if (existing >= 0) {
      _actors[existing] = entry;
    } else {
      // Newest first: the actor just cast is the one about to be used.
      _actors.insert(0, entry);
    }

    _persist();
    notifyListeners();
    return id;
  }

  Map<String, Object?> actor(String id) {
    for (final entry in _actors) {
      if ('${entry['id'] ?? ''}' == id) return Map<String, Object?>.of(entry);
    }
    return {};
  }

  void remove(String id) {
    final index = _actors.indexWhere((a) => '${a['id'] ?? ''}' == id);
    if (index < 0) return;

    final portrait = '${_actors[index]['portraitPath'] ?? ''}';
    _actors.removeAt(index);

    // Only files we adopted are ours to delete; a portrait the user pointed at
    // from their own folders stays where it is.
    if (portrait.isNotEmpty &&
        p.equals(p.dirname(File(portrait).absolute.path), _portraitDir)) {
      try {
        final file = File(portrait);
        if (file.existsSync()) file.deleteSync();
      } on FileSystemException {
        // Leaving an orphan file behind beats failing the delete.
      }
    }

    _persist();
    notifyListeners();
  }

  /// Suggests a name for an actor that has none, so saving never blocks on a
  /// dialog: "Actor 3".
  String suggestedName() => tr('Actor %1').arg(_actors.length + 1);

  static int _counter = 0;

  static String _newId() {
    _counter += 1;
    final stamp = DateTime.now().microsecondsSinceEpoch.toRadixString(16);
    return '${stamp.substring(stamp.length - 6)}$_counter';
  }
}
